### ISSUE TITLE
RubyGems is silent when trying to uninstall a gem outside of the gem home

### ISSUE THREAD

**Issue by eregon** [rubygems/rubygems#2701]
```
$ gem env                                                                     
RubyGems Environment:
  - RUBYGEMS VERSION: 3.0.3
  - RUBY VERSION: 2.6.2 (2019-03-13 patchlevel 47) [x86_64-linux]
  - INSTALLATION DIRECTORY: /home/eregon/.gem/ruby/2.6.2
  - USER INSTALLATION DIRECTORY: /home/eregon/.gem/ruby/2.6.0
  - RUBY EXECUTABLE: /home/eregon/.rubies/ruby-2.6.2/bin/ruby
  - GIT EXECUTABLE: /usr/bin/git
  - EXECUTABLE DIRECTORY: /home/eregon/.gem/ruby/2.6.2/bin
  - SPEC CACHE DIRECTORY: /home/eregon/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /home/eregon/.rubies/ruby-2.6.2/etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /home/eregon/.gem/ruby/2.6.2
     - /home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - "gem" => "--no-document"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /home/eregon/.gem/ruby/2.6.2/bin
     - /home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0/bin
     - /home/eregon/.rubies/ruby-2.6.2/bin
     - /home/eregon/bin
     - /usr/local/bin
     - /usr/local/sbin
     - /usr/bin
     - /usr/sbin
     - /home/eregon/code/mx
     - /home/eregon/code/mspec/bin
     - /opt/crystal/bin
$ gem install --install-dir /home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0 path
Successfully installed path-2.0.1
1 gem installed
$ gem uninstall path
# BUG: Nothing printed and it didn't remove the gem!
$ gem uninstall --debug path
NOTE:  Debugging mode prints all exceptions even when rescued
Exception `OptionParser::InvalidOption' at /home/eregon/.rubies/ruby-2.6.2/lib/ruby/2.6.0/optparse.rb:1769 - invalid option: no-document
Exception `OptionParser::InvalidOption' at /home/eregon/.rubies/ruby-2.6.2/lib/ruby/2.6.0/optparse.rb:1578 - invalid option: --no-document
Exception `Gem::InstallError' at /home/eregon/.rubies/ruby-2.6.2/lib/ruby/2.6.0/rubygems/uninstaller.rb:121 - path is not installed in GEM_HOME, try:
	gem uninstall -i /home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0 path
```

So `gem uninstall path` said nothing (and even exit code 0), even though it couldn't do what I asked (uninstalling the gem).
Running in `--debug` mode shows the message that I would expect to be printed by default.

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

Here are my current environment details:

```
$ gem env version
3.0.3
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

Relates to #2700.

**Comment by eregon** [comment]
Furthermore, it seems `gem uninstall -i some/dir path` also fails silently if `some/dir` is the correct directory but happens to be a symlink.

**Comment by bronzdoc** [comment]
Hi @eregon 

I'm tried to reproduce your problem but i can't, i tried with linux and mac and got the same results:

```shell
$gem install --install-dir /Users/bronzdoc/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0 nitv
Successfully installed nitv-0.1.1
Parsing documentation for nitv-0.1.1
Installing ri documentation for nitv-0.1.1
Done installing documentation for nitv after 0 seconds
1 gem installed
$ gem uninstall nitv
Successfully uninstalled nitv-0.1.1
```

in this particular case that should be the correct behavior, since i'm using `--install-dir` with a directory that is in my `GEM PATHS` as you are.

```
- GEM PATHS:
     - /Users/bronzdoc/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0
     - /Users/bronzdoc/.gem/ruby/2.5.0
```

> Furthermore, it seems gem uninstall -i some/dir path also fails silently if some/dir is the correct directory but happens to be a symlink (it works fine with the realpath of some/dir).

Please open a new issue for that.

**Comment by deivid-rodriguez** [comment]
I guess it's all related to symlinks?

**Comment by eregon** [comment]
@bronzdoc The first bug is about what `gem uninstall path` does, and as you can see above it outputs nothing, and exists with code 0, even though it didn't remove the gem installed in `/home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0` (not the gem home).

**Comment by deivid-rodriguez** [comment]
Yeah, I think the bug is valid, but we need a reproduction. I tried reproducing it and got a proper error, like "gem \`path\` is not installed", or something like that.

**Comment by eregon** [comment]
Second issue about symlinks split to #2705.

**Comment by eregon** [comment]
@deivid-rodriguez You should be able to reproduce with the commands I posted in the description, on a stock MRI 2.6.2, with the stock RubyGems.

**Comment by eregon** [comment]
I just tried on another machine, same results as in the description.

**Comment by deivid-rodriguez** [comment]
Let me try again.

**Comment by eregon** [comment]
@bronzdoc  Of course, you should install not to the first gem path, which is the same as gem home, but to another (e.g., the 2nd) gem path. Otherwise, it's just the same as not providing `--install-dir` at all.

**Comment by deivid-rodriguez** [comment]
Oh, I get it now, I was installing to any random folder that it's not `GEM_HOME`, I need to install to a `GEM_PATH` that it's not `GEM_HOME`. I'm pretty sure it will reproduce now :+1:.

**Comment by eregon** [comment]
Right, I didn't even think `gem` would let me install outside a gem path :D

**Comment by deivid-rodriguez** [comment]
It didn't, actually... :(

```
$ ruby -e 'puts Gem.paths.home' 
/home/deivid/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0

$ ruby -e 'puts Gem.paths.path' 
/home/deivid/.gem/ruby/2.6.0
/home/deivid/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0

$ gem install --install-dir /home/deivid/.gem/ruby/2.6.0 path
Fetching path-2.0.1.gem
Successfully installed path-2.0.1
1 gem installed

$ gem uninstall path
Successfully uninstalled path-2.0.1
```

**Comment by bronzdoc** [comment]
> @bronzdoc Of course, you should install not to the first gem path, which is the same as gem home, but to another (e.g., the 2nd) gem path. Otherwise, it's just the same as not providing --install-dir at all.

```
- GEM PATHS:
     - /Users/bronzdoc/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0
     - /Users/bronzdoc/.gem/ruby/2.5.0
```

```shell
$ gem install --install-dir /Users/bronzdoc/.gem/ruby/2.5.0 nitv
Successfully installed nitv-0.1.1
Parsing documentation for nitv-0.1.1
Installing ri documentation for nitv-0.1.1
Done installing documentation for nitv after 0 seconds
1 gem installed
$  gem uninstall nitv
Successfully uninstalled nitv-0.1.1
```

Still can't reproduce neither

**Comment by eregon** [comment]
Maybe this depends on `GEM_HOME`/`GEM_PATH` to be set in ENV, which rbenv doesn't set but chruby & RVM do.
Trying with either `chruby` or RVM should reproduce it

The CI reproduces it BTW: https://api.travis-ci.org/v3/job/513091459/log.txt
Look at the raw log:
```
$ gem uninstall executable-hooks gem-wrappers bundler-unload -x --force -i /home/travis/.rvm/gems/ruby-2.6.1@global
$ rake setup
# ^ no output at all, and the gems are still there:
...
bundler-unload (1.0.2)
    Author: Michal Papis
    Homepage: https://github.com/mpapis/bundler-unload
    License: Apache 2.0
    Installed at: /home/travis/.rvm/rubies/ruby-2.6.1/lib/ruby/gems/2.6.0

    Allow unloading bundler after Bundler.load
...
```

**Comment by deivid-rodriguez** [comment]
Yay! Reproduced under rvm :+1:

**Comment by bronzdoc** [comment]
Closed by https://github.com/rubygems/rubygems/pull/2707
